p2panda-rs/src/graph/graph.rs

```diff
@@ -0,0 +1,495 @@
+    }
+
+    /// Add a link between existing nodes to the graph. Returns true if the link was added.
+    /// Returns false if the link was unable to be added. This happens if either of the nodes were not
+    /// present in the graph, or if the link creates a single node loop.
+    pub fn add_link(&mut self, from: &str, to: &str) -> bool {
+        if from == to {
+            return false;
+        }
+
+        if let Some(from_node_mut) = self.0.get_mut(from) {
+            from_node_mut.next.push(to.to_owned());
+        } else {
+            return false;
+        }
+
+        if let Some(to_node_mut) = self.0.get_mut(to) {
+            to_node_mut.previous.push(from.to_owned());
+        } else {
+            return false;
+        }
+
+        true
+    }
+
+    /// Get node from the graph by key, returns `None` if it wasn't found.
+    pub fn get_node(&'a self, key: &str) -> Option<&Node<T>> {
+        self.0.get(key)
+    }
+
+    /// Get the data payload from this node.
+    pub fn get_node_data(&self, id: &str) -> Option<&T> {
+        self.0.get(id).map(|node| &node.data)
+    }
+
+    /// Returns true if this node key is connected to the graph.
+    pub fn is_connected(&self, key: &str) -> bool {
+        if let Some(node) = self.get_node(key) {
+            !node.previous().is_empty() || !node.next().is_empty()
+        } else {
+            false
+        }
+    }
+
+    /// Returns the keys for nodes which follows this node key.
+    pub fn get_next(&'a self, key: &str) -> Option<Vec<String>> {
+        self.get_node(key).map(|node| node.next())
+    }
+
+    /// Returns the keys for nodes which precede this node key.
+    pub fn get_previous(&'a self, key: &str) -> Option<Vec<String>> {
+        self.get_node(key).map(|node| node.previous())
+    }
+
+    /// Returns true if this node key is a merge node.
+    pub fn is_merge_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_merge(),
+            None => false,
+        }
+    }
+
+    /// Returns true if this node key is a branch node.
+    pub fn is_branch_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_branch(),
+            None => false,
+        }
+    }
+
+    /// Returns true if this node key is a graph tip.
+    pub fn is_tip_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_tip(),
+            None => false,
+        }
+    }
+
+    // NOT IMPLEMENTED //
+    // pub fn invalidate_keys(&self, keys: Vec<String>) {}
+
+    /// Returns a reference to the root node of this graph.
+    pub fn root_node(&self) -> &Node<T> {
+        self.0.values().find(|node| node.is_root()).unwrap()
+    }
+
+    /// Returns the root node key.
+    pub fn root_node_key(&self) -> String {
+        self.0.values().find(|node| node.is_root()).unwrap().key()
+    }
+
+    /// Check if all a nodes dependencies have been visited.
```

Comment:
Introducing the concept of dependencies here. Do we need it or is it the same as "previous nodes"?